modules/apstreamBidAdapter.js

```diff
@@ -0,0 +1,225 @@
.l^=r.l,n.l>>>=0}function a(n,r){var e={h:n.h<<r|n.l>>>32-r,l:n.l<<r|n.h>>>32-r};n.h=e.h,n.l=e.l}function u(n){var r=n.l;n.l=n.h,n.h=r}function i(n,r,e,t){o(n,r),o(e,t),a(r,13),a(t,16),x(r,n),x(t,e),u(n),o(e,r),o(n,t),a(r,17),a(t,21),x(r,e),x(t,n),u(e)}function l(n,r){return n[r+3]<<24|n[r+2]<<16|n[r+1]<<8|n[r]}function c(n,r){"string"==typeof r&&(r=function(n){if("function"==typeof TextEncoder)return(new TextEncoder).encode(n);n=unescape(encodeURIComponent(n));for(var r=new Uint8Array(n.length),e=0,t=n.length;e<t;e++)r[e]=n.charCodeAt(e);return r}(r));var e={h:n[1]>>>0,l:n[0]>>>0},t={h:n[3]>>>0,l:n[2]>>>0},o={h:e.h,l:e.l},a=e,u={h:t.h,l:t.l},c=t,f=r.length,h=f-7,s=new Uint8Array(new ArrayBuffer(8));x(o,{h:1936682341,l:1886610805}),x(u,{h:1685025377,l:1852075885}),x(a,{h:1819895653,l:1852142177}),x(c,{h:1952801890,l:2037671283});for(var d=0;d<h;){var v={h:l(r,d+4),l:l(r,d)};x(c,v),i(o,u,a,c),i(o,u,a,c),x(o,v),d+=8}s[7]=f;for(var p=0;d<f;)s[p++]=r[d++];for(;p<7;)s[p++]=0;var g={h:s[7]<<24|s[6]<<16|s[5]<<8|s[4],l:s[3]<<24|s[2]<<16|s[1]<<8|s[0]};x(c,g),i(o,u,a,c),i(o,u,a,c),x(o,g),x(a,{h:0,l:255}),i(o,u,a,c),i(o,u,a,c),i(o,u,a,c),i(o,u,a,c);var y=o;return x(y,u),x(y,a),x(y,c),y}var f=function(n,r){var e=c(n,r);return("0000000"+e.h.toString(16)).substr(-8)+("0000000"+e.l.toString(16)).substr(-8)}.bind(null,[2251905623,1801168887,1769155591,711914637]),h=new RegExp("^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?"),s=null,d={};function v(){var n=s||window.location.toString();if(n=n.replace(/\.demo\.audienceproject\.com\//,"/"),d.url===n)return d.parsed;var r=function(n){var r=0!==n.indexOf("/")&&-1!==n.indexOf("/")&&(-1===n.indexOf(":")||n.indexOf(":")>n.indexOf("/")),e=h.exec(r?"noscheme://"+n:n),t={scheme:r?"":e[2]||"",host:e[4]||"",hostname:e[4]?e[4].split(":")[0]:"",pathname:e[5]||"",search:e[7]||"",hash:e[9]||"",toString:function(){return n}};return t.origin=t.scheme+"://"+t.host,t}(n);return d.url=n,d.parsed=r,r}function p(){var n=function(){if("function"==typeof Uint32Array&&"undefined"!=typeof crypto&&void 0!==crypto.getRandomValues){var n=new Uint32Array(4);crypto.getRandomValues(n);var r=-1;return"xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g,function(e){var t=n[++r>>3]>>r%8*4&15;return("x"===e?t:3&t|8).toString(16)})}var e=(new Date).getTime();return"undefined"!=typeof performance&&"function"==typeof performance.now&&(e+=performance.now()),"xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g,function(n){var r=(e+16*Math.random())%16|0;return e=Math.floor(e/16),("x"===n?r:3&r|8).toString(16)})}(),o=v(),x=f(n+o.toString()),a=x.substr(0,4),u=x.substr(4);n=n.substr(0,19)+a+"-"+u;var i,l=(i=(new Date).getTime()-new Date(2019,0,1).getTime(),Math.floor(i/864e5)),c=f(o.origin),h=[t.USERREPORT,l,c].join("."),s=f(n+h+e);return[r,s,n,h].join(".")}return{readOrCreateDsu:function(){var r;try{r=storage.getDataFromLocalStorage(n)}catch(n){return null}r||(r=p());try{storage.setDataInLocalStorage(n,r)}catch(n){return null}return r}}}();
```

Comment:
What is this? This doesn't look transparent, I can't know what side effects this code has, to the page or to the pb.js auction.

Reply:
This is our internal module to generate unique token with some metadata and store into local storage. It does not interact with anything except local storage.

Reply:
You can use some of the built-in functions for generating unique identifiers (like generateUUID). Also, you should be using Prebid methods for working with local storage (look for storageManager), this way you get GDPR enforcement rules out of the box.

Reply:
I'v added one more commit where our module is re-wrote in the clear way. Also I realized what we are using standard UUID generation so it has been replaced with Prebid's generateUUID implementation. And for working with the local storage I used Prebid's methods.

Reply:
this looks much better, thank you